### PR TITLE
Issue#422 Implementation of CreateCluster() API

### PR DESCRIPTION
**Pull request by roopak-qlogic**
Implementation of CreateCluster API will be part of this PR.
Unit test case, Integration test and Sample example will be part of different PR

### CODE REVIEW

File: bigtable/client/instance_admin.cc [hunk 1]
```diff
@@ -108,6 +108,80 @@ google::bigtable::admin::v2::Instance InstanceAdmin::CreateInstanceImpl(
...
 }
 
+std::future<google::bigtable::admin::v2::Cluster> InstanceAdmin::CreateCluster(
+    ClusterConfig instance_config, std::string const& instance_id,
+    std::string const& cluster_id) {
+  return std::async(std::launch::async, &InstanceAdmin::CreateClusterImpl, this,
+                    std::move(instance_config), instance_id, cluster_id);
+}
+
+google::bigtable::admin::v2::Cluster InstanceAdmin::CreateClusterImpl(
```
Comment (coryan) [inline bigtable/client/instance_admin.cc]: Let's keep the functions in the same order they are declared in the header file.

File: bigtable/client/instance_admin.cc [hunk 2]
```diff
@@ -108,6 +108,80 @@ google::bigtable::admin::v2::Instance InstanceAdmin::CreateInstanceImpl(
...
+    ClusterConfig const& cluster_config, std::string const& instance_id,
+    std::string const& cluster_id) {
```
Comment (coryan) [inline bigtable/client/instance_admin.cc]: Please use strong types for `instance_id` and `cluster_id`.  I believe other functions can serve as examples.

File: bigtable/client/instance_admin.h [hunk 3]
```diff
@@ -78,6 +78,12 @@ class InstanceAdmin {
   std::future<google::bigtable::admin::v2::Instance> CreateInstance(
       InstanceConfig instance_config);
 
+  /// Create Cluster.
```
Comment (coryan) [inline bigtable/client/instance_admin.h]: This Doxygen comment is super minimal.

File: bigtable/client/instance_admin.cc [hunk 4]
```diff
@@ -108,6 +108,80 @@ google::bigtable::admin::v2::Instance InstanceAdmin::CreateInstanceImpl(
...
+      impl_.metadata_update_policy_, &InstanceAdminClient::CreateCluster,
+      request, "InstanceAdmin::CreateCluster", status, false);
+  if (not status.ok()) {
+    bigtable::internal::RaiseRpcError(status,
+                                      "unrecoverable error in MakeCall()");
+  }
+
+  google::bigtable::admin::v2::Cluster result;
+
+  while (not response.done()) {
```
Comment (coryan) [inline bigtable/client/instance_admin.cc]: Unfortunately my implementation of `CreateInstance` had a bug which you reproduced here: if the longrunning operation is immediately satisfied (i.e. if `response.done() == true` the first time you enter this loop), the `result` variable is never initialized.  A recent merge has a new implementation without the bug, please fix.  We should consider refactoring too.

File: bigtable/client/instance_admin.h [hunk 5]
```diff
@@ -78,6 +78,21 @@ class InstanceAdmin {
   std::future<google::bigtable::admin::v2::Instance> CreateInstance(
       InstanceConfig instance_config);
 
+  /**
+     * Create a new Cluster of Cloud Bigtable..
```
Comment (coryan) [inline bigtable/client/instance_admin.h]: Plesae fix indentation.  Also remove double period at the end of this sentence.

**Comment by roopak-qlogic** [comment]
updated PR with all the changes from the master

**Comment by roopak-qlogic** [comment]
I have written the Unit test case for create cluster for "no error", but some how I am getting backtrace in my unit test case "CreateCluster"  on line number 721 : auto actual = future.get();
I tried and debug but do not know where i am missing, and did not get any clue yet, If I get any pointer on this.
I have saved my code in below feature branch
https://github.com/GoogleCloudPlatform/google-cloud-cpp/compare/master...roopak-qlogic:Issue-422_SAVE?expand=1

**Comment by roopak-qlogic** [comment]
Backtrace: error has been resolved, Thanks

File: bigtable/client/instance_admin.cc [hunk 6]
```diff
@@ -147,5 +154,72 @@ void InstanceAdmin::DeleteCluster(bigtable::InstanceId const& instance_id,
...
+      if (response.has_error()) {
+        bigtable::internal::RaiseRpcError(
+            grpc::Status(static_cast<grpc::StatusCode>(response.error().code()),
+                         response.error().message()),
+            "long running op failed");
+      }
+    }
+    // Wait before polling, and then poll the operation to get the new
+    // "response.
+    auto delay = backoff_policy->on_completion(status);
```
Comment (coryan) [inline bigtable/client/instance_admin.cc]: Please add a TODO entry saying that we should use the PollingPolicy here once #461 is merged.

File: bigtable/client/instance_admin.h (hunk 5 above)
Comment (coryan) [reply bigtable/client/instance_admin.h]: This is still a problem btw.

**Comment by roopak-qlogic** [comment]
yes I am doing my changes in my one of the branch, just for save my code, once they complete I will include in this PR for review, even bigtable/client/instance_admin.h indentation also fixed in my saved copy branch, tomorrow I will push all for review

**Comment by roopak-qlogic** [comment]
Integration test and sample snippet will be in separate PR

File: bigtable/client/instance_admin_test.cc [hunk 7]
```diff
@@ -930,3 +930,165 @@ TEST_F(InstanceAdminTest, DeleteClusterRecoverableError) {
...
+                          btproto::CreateClusterRequest const& request,
+                          google::longrunning::Operation* response) {
+        auto const project_name =
+            "projects/" + kProjectId + "/instances/test-instance";
+        EXPECT_EQ(project_name, request.parent());
+        return grpc::Status::OK;
+      }));
+
+  std::string expected_text = R"(
+name: 'projects/my-project/instances/test-instance'
```
Comment (coryan) [inline bigtable/client/instance_admin_test.cc]: Please indent, like so:

```C++
  std::string expected_text = R"(
      name: 'projects/my-project/instances/test-instance'
```

File: bigtable/client/instance_admin_test.cc [hunk 8]
```diff
@@ -930,3 +930,165 @@ TEST_F(InstanceAdminTest, DeleteClusterRecoverableError) {
...
+location: 'Location1'
+default_storage_type: SSD
+)";
+  btproto::Cluster expected;
+  ASSERT_TRUE(
+      google::protobuf::TextFormat::ParseFromString(expected_text, &expected));
+  EXPECT_CALL(*client_, GetOperation(_, _, _))
+      .WillOnce(Invoke([](grpc::ClientContext*,
```
Comment (coryan) [inline bigtable/client/instance_admin_test.cc]: Consider storing the lambda for these two first cases in a variable, and reduce the clutter a bit.

File: bigtable/client/instance_admin_test.cc [hunk 9]
```diff
@@ -930,3 +930,165 @@ TEST_F(InstanceAdminTest, DeleteClusterRecoverableError) {
...
+
+/// @test Verify that `bigtable::InstanceAdmin::CreateCluster` works.
+TEST_F(InstanceAdminTest, CreateClusterImmediatelyReady) {
+  using ::testing::_;
+  using ::testing::Invoke;
+
+  bigtable::InstanceAdmin tested(client_);
+
+  std::string expected_text = R"(
+name: 'projects/my-project/instances/test-instance'
```
Comment (coryan) [inline bigtable/client/instance_admin_test.cc]: Indent as described before.

File: bigtable/client/instance_admin_test.cc [hunk 10]
```diff
@@ -930,3 +930,165 @@ TEST_F(InstanceAdminTest, DeleteClusterRecoverableError) {
...
+                          btproto::CreateClusterRequest const& request,
+                          google::longrunning::Operation* response) {
+        auto const project_name =
+            "projects/" + kProjectId + "/instances/test-instance";
+        EXPECT_EQ(project_name, request.parent());
+        return grpc::Status::OK;
+      }));
+
+  std::string expected_text = R"(
+name: 'projects/my-project/instances/test-instance'
```
Comment (coryan) [inline bigtable/client/instance_admin_test.cc]: Indent as described before.

File: bigtable/client/instance_admin_test.cc [hunk 11]
```diff
@@ -930,3 +930,165 @@ TEST_F(InstanceAdminTest, DeleteClusterRecoverableError) {
...
+location: 'Location1'
+default_storage_type: SSD
+)";
+  btproto::Cluster expected;
+  ASSERT_TRUE(
+      google::protobuf::TextFormat::ParseFromString(expected_text, &expected));
+  EXPECT_CALL(*client_, GetOperation(_, _, _))
+      .WillOnce(Invoke([](grpc::ClientContext*,
```
Comment (coryan) [inline bigtable/client/instance_admin_test.cc]: Consider storing the lambda in a variable to reduce code clutter.

File: bigtable/client/instance_admin.h [hunk 12]
```diff
@@ -84,6 +84,20 @@ class InstanceAdmin {
...
+   * @param cluster_id the id of the cluster in the project that needs to be
+   *   created
+   *
+   *  @par Example
+   *  TODO(#422) implement tests and examples for CreateCluster
+   */
+  std::future<google::bigtable::admin::v2::Cluster> CreateCluster(
+      ClusterConfig cluster_config, bigtable::InstanceId const& instance_id,
+      bigtable::ClusterId const& cluster_id);
   /**
```
Comment (coryan) [inline bigtable/client/instance_admin.h]: Please leave a blank line here.

File: bigtable/client/instance_admin.h [hunk 13]
```diff
@@ -105,7 +119,6 @@ class InstanceAdmin {
    */
   std::future<google::bigtable::admin::v2::Instance> UpdateInstance(
       InstanceUpdateConfig instance_update_config);
-
```
Comment (coryan) [inline bigtable/client/instance_admin.h]: Please do not remove this blank line.

**Comment by roopak-qlogic** [comment]
Updated, Now it should be OK.

**Pull request merged**